backend/controller/ml_model/compiler.py

```diff
@@ -81,77 +86,96 @@ class UniqueId(shared.StrEnum):
-    self.source = source
-    self.input = input
-    self.features = []
-    self.label = None
-    self.trigger_event = None
-    self.first_value = None
-    self._trigger_date = None
-    self._unique_id = unique_id.lower()
-    self.user_id = 'user_id'
-    self.client_id = 'user_pseudo_id'
+  _items: list[models.MlModelVariable]
+  _source: Source
+  _input: models.MlModelInput
+  _unique_id: UniqueId
+
+  def __init__(self,
+               source: Source,
+               ml_model: models.MlModel) -> None:
+    self._items = []
+    for variable in ml_model.variables:
+      if Source(variable.source) == source:
+        self._items.append(variable)
+
+    self._source = source
+    self._input = ml_model.input
+    self._unique_id = ml_model.unique_id
 
   @property
   def in_source(self) -> bool:
-    return self.source in self.input.source
+    return self._source in self._input.source
 
   @property
   def dataset(self):
-    if self.input.parameters:
-      return self.input.parameters.first_party_dataset
+    if self._input.parameters:
+      return self._input.parameters.first_party_dataset
 
   @property
   def table(self):
-    if self.input.parameters:
-      return self.input.parameters.first_party_table
+    if self._input.parameters:
+      return self._input.parameters.first_party_table
 
   @property
   def unique_id(self):
-    unique_id = self.__getattribute__(self._unique_id)
-    return unique_id.name if isinstance(unique_id, models.MlModelVariable) else unique_id
+    unique_id = self._single(self._unique_id)
+    default = 'user_id' if self._unique_id == UniqueId.USER_ID else 'user_pseudo_id'
+    return unique_id if unique_id else {'name': default}
+
+  @property
+  def features(self):
+    return self._list(VariableRole.FEATURE)
+
+  @property
+  def label(self):
+    return self._single(VariableRole.LABEL)
+
+  @property
+  def first_value(self):
+    return self._single(VariableRole.FIRST_VALUE)
+
+  @property
+  def trigger_event(self):
+    return self._single(VariableRole.TRIGGER_EVENT)
 
   @property
   def trigger_date(self):
-    if self.source == Source.FIRST_PARTY:
-      return self._trigger_date
-    elif self.source == Source.GOOGLE_ANALYTICS:
+    if self._source == Source.FIRST_PARTY:
+      return self._single(VariableRole.TRIGGER_DATE)
+    elif self._source == Source.GOOGLE_ANALYTICS:
       if self.trigger_event:
         return self.trigger_event
       elif self.first_value:
         return self.first_value
       else:
         return self.label
 
-  def add(self, variable: models.MlModelVariable) -> None:
-    if variable.role == VariableRole.FEATURE:
-      self.features.append(variable)
-    elif variable.role == VariableRole.TRIGGER_DATE:
-      self._trigger_date = variable
-    else:
-      self.__setattr__(str(variable.role).lower(), variable)
+  @property
+  def gclid(self):
+    return self._single(VariableRole.GCLID)
+
+  def _single(self, role: VariableRole) -> models.MlModelVariable:
```

Comment:
Please add a comment on why this function returns ```None```` if there is only one model variable.

Reply:
Ah, my mistake.  Closing conversation.